### PR TITLE
Adding backtranslation expander

### PR DESCRIPTION
**Pull request by DelaramRajaei**
- Reformating the code
- Add the backtranslation expander 
- Details are mentioned in this [issue](https://github.com/fani-lab/ReQue/issues/27)

### CODE REVIEW

**Comment by hosseinfani** [comment]
Hi @DelaramRajaei 
Thank you for the nice changes. I put some comments. Please address them. Also, attach a script or sth in the issue page that shows the smooth run of the pipeline with added features as a unit/integration test.

**Comment by DelaramRajaei** [comment]
Hi @hosseinfani 
I updated some comments I remembered but unfortunately, I can not see the comments you put on the code.
Also, I update the return of all the expanders. In this new update, the semsim score is stored in both txt and csv files.

**Comment by hosseinfani** [comment]
@DelaramRajaei 
I had to submit my comments. Now you can see them I believe.

File: qe/main.py [hunk 1]
```diff
@@ -1,140 +1,138 @@
...
         for ranker in rankers:
-            Q_pred = '{}.{}.{}.txt'.format(output, model_name, utils.get_ranker_name(ranker))
+            Q_pred = f'{output}.{model_name}.{utils.get_ranker_name(ranker)}.txt'
             for metric in metrics:
-                Q_eval = '{}.{}.{}.{}.txt'.format(output, model_name, utils.get_ranker_name(ranker), metric)
-                cli_cmd = '\"{}\" -q -m {} {} {} > {}'.format(eval_cmd, metric, Qrels, Q_pred, Q_eval)
-                print('{}\n'.format(cli_cmd))
+                Q_eval = f'{output}.{model_name}.{utils.get_ranker_name(ranker)}.{metric}.txt'
+                cli_cmd = f'\"{eval_cmd}\" -q -m {metric} {Qrels} {Q_pred} > {Q_eval}'
+                print(f'{cli_cmd}\n')
```
Comment (hosseinfani) [inline qe/main.py:130]: print(cli_cmd)

File: qe/main.py [hunk 2]
```diff
@@ -143,251 +141,155 @@ def aggregate(expanders, rankers, metrics, output):
...
         #     model_errs[model_name] = traceback.format_exc()
         #     continue
         df = pd.concat([df, Q_], axis=1)
 
-    filename = '{}.{}.{}.all.csv'.format(output, '.'.join([utils.get_ranker_name(r) for r in rankers]), '.'.join(metrics))
+    filename = f"{output}.{'.'.join([utils.get_ranker_name(r) for r in rankers])}.{'.'.join(metrics)}.all.csv"
     df.to_csv(filename, index=False)
     # for model_err, msg in model_errs.items():
-    #     print('INFO: MAIN: AGGREGATE: There has been error in {}!\n{}'.format(model_err, msg))
+    #     print(f'INFO: MAIN: AGGREGATE: There has been error in {model_err}!\n{msg}')
```
Comment (hosseinfani) [inline qe/main.py:161]: @DelaramRajaei 
I liked that you also made the changes to the commented area. It's forward-thinking! thank you.

File: qe/main.py [hunk 3]
```diff
@@ -143,251 +141,155 @@ def aggregate(expanders, rankers, metrics, output):
...
+            if 'evaluate' in op: evaluate(expander=expander, Qrels=param.corpora[corpus]['qrels'], rankers=rankers,
+                                          metrics=metrics, eval_cmd=param.anserini['trec_eval'], output=output_)
         except:
             print(f'INFO: MAIN: THREAD: {threading.currentThread().getName()}: There has been error in {expander}!\n{traceback.format_exc()}')
             exceptions[expander.get_model_name()] = traceback.format_exc()
 
     threads = []
     for expander in expanders:
-        if param.ReQue['parallel']: threads.append(threading.Thread(daemon=True, target=worker_thread, name=expander.get_model_name(), args=(expander,)))
+        if param.ReQue['parallel']:
```
Comment (hosseinfani) [inline qe/main.py:219]: @DelaramRajaei 
make it one line pls

File: qe/main.py [hunk 4]
```diff
@@ -143,251 +141,155 @@ def aggregate(expanders, rankers, metrics, output):
...
+    corpus.add_argument('--corpus', type=str,
+                        choices=['dbpedia', 'antique', 'robust04', 'gov2', 'clueweb09b', 'clueweb12b13', 'trec09mq',
+                                 'orcas', 'testds'], required=True,
+                        help='The corpus name; required; (example: robust04)')
 
     gold = parser.add_argument_group('Gold Standard Dataset')
-    gold.add_argument('--output', type=str, required=True, help='The output path for the gold standard dataset; required; (example: ./output/robust04/')
-    gold.add_argument('--rankers', nargs='+', type=str.lower, choices=['bm25', 'qld'], default=['bm25', 'qld'], help='The ranker names (default: bm25 qld)')
-    gold.add_argument('--metrics', nargs='+', type=str.lower, choices=['map', 'ndcg'], default=['map', 'ndcg'], help='The evaluation metric names (default: map ndcg)')
+    gold.add_argument('--output', type=str, required=True,
```
Comment (hosseinfani) [inline qe/main.py:286]: @DelaramRajaei 
make them one line pls

**Comment by DelaramRajaei** [comment]
@hosseinfani 
I went through the comments and implemented the suggested changes. 
I pushed the updated version.

**Comment by hosseinfani** [comment]
> Hi @hosseinfani I updated some comments I remembered but unfortunately, I can not see the comments you put on the code. Also, I update the return of all the expanders. In this new update, the semsim score is stored in both txt and csv files.
> 
> > Also, attach a script or sth in the issue page that shows the smooth run of the pipeline with added features as a unit/integration test.
> 
> Could you please explain this a little? Can I attach the csv file as results?

@DelaramRajaei 
let me when you're available today (check msteams). I'll explain more then.

**Comment by hosseinfani** [comment]
@DelaramRajaei 
just a quick check, if an expander is not a backtranslated one, there is no problem with semsim value?

**Pull request merged**